File: src/net/merayen/elastic/ui/objects/components/CircularSlider.java
package net.merayen.elastic.ui.objects.components;

import net.merayen.elastic.ui.Draw;
import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.event.UIEvent;
import net.merayen.elastic.ui.util.MouseHandler;
import net.merayen.elastic.util.Point;

public class CircularSlider extends UIObject {
	public interface Handler {
		public void onChange(float value);
	}

	public float size = 30f;
	public float drag_scale = 0.25f;
	public float pointer_length = 1f;

	// In radian, min and max position
	public float min = (float) Math.PI*1.8f;
	public float max = (float) Math.PI*0.2f;

	private float value = 0;

	private MouseHandler mousehandler;
	private float drag_value;

	private Handler handler;

	public void onInit() {
		mousehandler = new MouseHandler(this);
		mousehandler.setHandler(new MouseHandler.Handler() {
			@Override
			public void onMouseDrag(Point start_point, Point offset) {
				setValue(drag_value - offset.y / (size / drag_scale));
				if(handler != null)
					handler.onChange(value);
			}

			@Override
			public void onMouseDown(Point position) {
				drag_value = value;
			}
		});
	}

	@Override
	public void onDraw(Draw draw) {
		draw.setStroke(size / 20);
		draw.setColor(30, 30, 30);
		draw.oval(0, 0, size, size);

		draw.setColor(30, 30, 30);
		drawLine(draw, 0, 0.8f);
		drawLine(draw, 1, 0.8f);

		draw.setColor(200, 200, 200);
		//draw.setStroke(size / 10);
		drawLine(draw, value, 0.6f);
	}

	@Override
	public void onEvent(UIEvent e) {
		mousehandler.handle(e);
	}

	private void drawLine(Draw draw, float value, float length) {
		//value = Math.max(Math.min(value, 1f), 0f);
		draw.line(
			size/2 + (float)Math.sin(min + value * (max-min)) * (size * length) / 2.1f,
			size/2 + (float)Math.cos(min + value * (max-min)) * (size * length) / 2.1f,
			size/2 + (float)Math.sin(min + value * (max-min)) * size / 2.3f,
			size/2 + (float)Math.cos(min + value * (max-min)) * size / 2.3f
		);
	}

	public void setValue(float value) {
		this.value = Math.min(Math.max(value, 0f), 1f);
	}

	public float getValue() {
		return value;
	}

	public void setHandler(Handler handler) {
		this.handler = handler;
	}
}

File: src/net/merayen/elastic/ui/objects/components/CircularSlider.kt
package net.merayen.elastic.ui.objects.components

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.util.MouseHandler
import net.merayen.elastic.util.Point

class CircularSlider : UIObject() {
	var size = 30f
	var dragScale = 0.25f
	var handler: Handler? = null

	val label = Label()
	val valueLabel = Label()

	// In radian, min and max position
	var min = Math.PI.toFloat() * 1.8f
	var max = Math.PI.toFloat() * 0.2f

	var value = 0f
		set(value) {
			if(value != field) {
				field = Math.min(Math.max(value, 0f), 1f)
				valueLabel.text = handler?.onLabelUpdate(field) ?: ""
				println("Value change ${field} ${this.value} ${handler}")
			}
		}

	lateinit var mousehandler: MouseHandler
	private var dragValue: Float = 0.toFloat()

	interface Handler {
		fun onChange(value: Float)
		fun onLabelUpdate(value: Float) = ""
	}

	override fun onInit() {
		val mousehandler = MouseHandler(this)
		mousehandler.setHandler(object : MouseHandler.Handler() {
			override fun onMouseDrag(start_point: Point, offset: Point) {
				val prevValue = value
				value = dragValue - offset.y / (size / dragScale)
				if(value != prevValue) {
					handler?.onChange(value)
					valueLabel.text = handler?.onLabelUpdate(value) ?: ""
				}
			}

			override fun onMouseDown(position: Point) {
				dragValue = value
			}
		})

		this.mousehandler = mousehandler

		label.align = Label.Align.CENTER
		label.eventTransparent = true
		add(label)

		valueLabel.align = Label.Align.CENTER
		valueLabel.eventTransparent = true
		add(valueLabel)
	}

	override fun onDraw(draw: Draw) {
		draw.setStroke(1.5f)
		draw.setColor(30, 30, 30)
		draw.oval(0f, 0f, size, size)

		draw.setColor(30, 30, 30)
		drawLine(draw, 0f, 0.8f)
		drawLine(draw, 1f, 0.8f)

		draw.setColor(200, 200, 200)
		drawLine(draw, value, 0.6f)
	}

	override fun onUpdate() {
		label.translation.x = size / 2
		label.translation.y = size
		label.fontSize = size / 4

		valueLabel.translation.x = size / 2
		valueLabel.translation.y = size / 3
		valueLabel.fontSize = size / 5
	}

	override fun onEvent(e: UIEvent) {
		mousehandler.handle(e)
	}

	private fun drawLine(draw: Draw, value: Float, length: Float) {
		//value = Math.max(Math.min(value, 1f), 0f);
		draw.line(
				size / 2 + Math.sin((min + value * (max - min)).toDouble()).toFloat() * (size * length) / 2.1f,
				size / 2 + Math.cos((min + value * (max - min)).toDouble()).toFloat() * (size * length) / 2.1f,
				size / 2 + Math.sin((min + value * (max - min)).toDouble()).toFloat() * size / 2.3f,
				size / 2 + Math.cos((min + value * (max - min)).toDouble()).toFloat() * size / 2.3f
		)
	}
}
